### PR TITLE
New channel manager

### PR DESCRIPTION
**Pull request by ArtoLord**

### CODE REVIEW

File: lzy/lzy-service/src/main/resources/log4j2.yaml [hunk 1]
```diff
@@ -3,7 +3,7 @@ Configuration:
   properties:
     property:
       name: gc_filename
-      value: target/garbage_collector.log
+      value: peer/garbage_collector.log
```
Comment (imakunin) [inline lzy/lzy-service/src/main/resources/log4j2.yaml:6]: really? :)

File: test/src/test/java/ai/lzy/test/impl/v2/AllocatorContext.java [hunk 2]
```diff
@@ -94,7 +94,7 @@ public static class WorkerAllocatorContext extends AllocatorContext {
 
         @Inject
         public WorkerAllocatorContext(IamContext iam) {
-            super(iam, "../lzy/worker/target/worker.jar", "ai.lzy.worker.Worker", 23910);
+            super(iam, "../lzy/worker/peer/worker.jar", "ai.lzy.worker.Worker", 23910);
```
Comment (imakunin) [inline test/src/test/java/ai/lzy/test/impl/v2/AllocatorContext.java:97]: and here :)

File: lzy/channel-manager/src/main/resources/db/channel-manager/migrations/v2/V1_0__Init_tables.sql [hunk 3]
```diff
@@ -0,0 +1,33 @@
+CREATE TABLE channel
```
Comment (imakunin) [inline lzy/channel-manager/src/main/resources/db/channel-manager/migrations/v2/V1_0__Init_tables.sql:1]: channels

File: lzy/channel-manager/src/main/resources/db/channel-manager/migrations/v2/V1_0__Init_tables.sql [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    id                   text      NOT NULL PRIMARY KEY,
+    owner_id             text      NOT NULL,
+    execution_id         text      NOT NULL,
+    workflow_name        text      NOT NULL,
+    data_scheme_json     text      NULL,
+    storage_producer_uri text      NULL,
+    storage_consumer_uri text      NULL
+);
+
+CREATE TABLE peer
```
Comment (imakunin) [inline lzy/channel-manager/src/main/resources/db/channel-manager/migrations/v2/V1_0__Init_tables.sql:12]: peers

File: lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/ActionScheduler.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+            throw new RuntimeException(e);
+        }
+
+        for (var transmission: transmissions) {
+            scheduleStartTransferAction(transmission.slot(), transmission.peer());
+        }
+    }
+
+    private void run(Peer slot, Peer peer) {
+
```
Comment (imakunin) [inline lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/ActionScheduler.java:56]: redundant empty line

File: lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/ActionScheduler.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+            LOG.error("Cannot restore pending transmissions", e);
+            throw new RuntimeException(e);
+        }
+
+        for (var transmission: transmissions) {
+            scheduleStartTransferAction(transmission.slot(), transmission.peer());
+        }
+    }
+
+    private void run(Peer slot, Peer peer) {
```
Comment (imakunin) [inline lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/ActionScheduler.java:55]: the meaning of the arguments is unclear

Comment (imakunin) [inline lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/ActionScheduler.java:55]: add logs, plz

File: lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/ActionScheduler.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+            connection.v2SlotsApi().startTransfer(
+                LSA.StartTransferRequest.newBuilder()
+                    .setSlotId(peer.id())
+                    .setPeer(slot.peerDescription())
+                    .build()
+            );
+
+            DbHelper.withRetries(LOG, () -> connections.dropPendingTransmission(slot.id(), peer.id(), null));
+        } catch (Exception e) {
+
```
Comment (imakunin) [inline lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/ActionScheduler.java:73]: redundant empty line

File: lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/db/ChannelDaoImpl.java [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+    @Override
+    public Channel create(String id, String userId, String executionId, String workflowName,
+                          LMD.DataScheme dataScheme, String storageProducerUri, String storageConsumerUri,
+                          TransactionHandle tx) throws SQLException
+    {
+        return DbOperation.execute(tx, storage, connection -> {
+            try (PreparedStatement ps = connection.prepareStatement("""
+                INSERT INTO channels (%s)
+                VALUES (?, ?, ?, ?, ?, ?, ?)
+                """.formatted(FIELDS)))
```
Comment (imakunin) [inline lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/db/ChannelDaoImpl.java:49]: introduce new constant fields for this query, plz

File: lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/db/ChannelDaoImpl.java [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+@Singleton
+public class ChannelDaoImpl implements ChannelDao {
+    private static final String FIELDS = "id, owner_id, execution_id, workflow_name, data_scheme_json," +
+        " storage_producer_uri, storage_consumer_uri";
+
+    private static final String CHANNEL_FIELDS = "channels.id, channels.owner_id, channels.execution_id, " +
```
Comment (imakunin) [inline lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/db/ChannelDaoImpl.java:28]: use multiline string format, plz

File: lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/db/ChannelDaoImpl.java [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+        var peers = (String[]) rs.getArray(8).getArray();
+        final List<LC.PeerDescription> producers = new ArrayList<>();
+        final List<LC.PeerDescription> consumers = new ArrayList<>();
+
+        for (var peer: peers) {
+            if (peer == null) {  // No peers for this channel
+                continue;
+            }
+
+            var parts = peer.split("@@@");
```
Comment (imakunin) [inline lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/db/ChannelDaoImpl.java:227]: what's the heck? :)

File: lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/db/PeerDao.java [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+public interface PeerDao {
+
+    Peer create(String channelId, LC.PeerDescription desc, Peer.Role role, Priority priority, boolean connected,
+                TransactionHandle tx) throws SQLException;
+
+    /**
+     * Find most prior producer in this channel
+     */
+    @Nullable
+    Peer findPriorProducer(String channelId, TransactionHandle tx) throws SQLException;
```
Comment (imakunin) [inline lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/db/PeerDao.java:20]: s/prior/prev (or previous)

File: lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/ChannelManagerMain.java [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+        final var authInterceptor = new AuthServerInterceptor(
+            new AuthenticateServiceGrpcClient(SERVICE_NAME, iamChannel));
+
+        final var channelManagerAddress = HostAndPort.fromString(config.getAddress());
+
+        final var internalOnly = new AllowInternalUserOnlyInterceptor(SERVICE_NAME, iamChannel);
+
+        this.server = newGrpcServer(channelManagerAddress, authInterceptor)
+            .addService(slotsService)
```
[the remark this answers is not in the document]
Comment (ArtoLord) [reply lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/ChannelManagerMain.java:48]: There is `authInterceptor`.  It adds global auth interceptor to all methods.
We check access already inside the method

File: lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/ActionScheduler.java [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+    public ActionScheduler(TransferDao connections, SlotConnectionManager connectionManager,
+                           ChannelOperationExecutor operationExecutor, Utils utils)
+    {
+        this.connections = connections;
+        this.connectionManager = connectionManager;
+        this.operationExecutor = operationExecutor;
+        this.utils = utils;
+    }
+
+    public void scheduleStartTransferAction(Peer slot, Peer peer) {
```
Comment (imakunin) [inline lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/ActionScheduler.java:37]: not `schedule` but `run`
`schedule` requires delay period

File: lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/Utils.java [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+import io.grpc.ManagedChannel;
+import jakarta.inject.Named;
+import jakarta.inject.Singleton;
+
+import java.sql.SQLException;
+
+import static ai.lzy.util.grpc.GrpcUtils.newBlockingClient;
+
+@Singleton
+public class Utils {
```
Comment (imakunin) [inline lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/Utils.java:18]: ambigous name for instantiated class

File: docs/tutorials/3-basics.md [hunk 15]
```diff
@@ -7,7 +7,7 @@ annotations are required.** For example:
 
 ```python
 from sklearn import datasets
-from sklearn.utils import Bunch
+from sklearn.lzyServiceClient import Bunch
```
Comment (imakunin) [inline docs/tutorials/3-basics.md:10]: why?

File: lzy/channel-manager/src/main/java/ai/lzy/channelmanager/operation/action/ChannelAction.java [hunk 16]
```diff
@@ -277,7 +277,7 @@ protected void disconnect(Endpoint endpoint) {
             if (Status.NOT_FOUND.getCode().equals(e.getStatus().getCode()) ||
                 Status.UNAVAILABLE.getCode().equals(e.getStatus().getCode()))
             {
-                LOG.info("Async operation (operationId={}): disconnectSlot request failed, slot not found. "
+                LOG.info("Async operation (operationId={}): disconnectSlot request failed, from not found. "
```
Comment (imakunin) [inline lzy/channel-manager/src/main/java/ai/lzy/channelmanager/operation/action/ChannelAction.java:280]: one more artefact after `mass replace`

File: lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/services/ChannelService.java [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+        } else {
+            LOG.error("{} Consumer and producer not set", logPrefix);
+
+            throw Status.INVALID_ARGUMENT
+                .withDescription("Consumer or producer must be set")
+                .asRuntimeException();
+        }
+
+
+        var peerId = idGenerator.generate("storage_peer-");
```
Comment (imakunin) [inline lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/services/ChannelService.java:60]: `peer` or `storage-peer` ?

File: lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/services/ChannelService.java [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+        };
+
+        try {
+            var res = DbHelper.withRetries(LOG, () -> {
+                try (var tx = TransactionHandle.create(storage)) {
+                    var channel = channelDao.find(request.getUserId(), request.getExecutionId(), storageProducerUri,
+                        storageConsumerUri, tx);
+
+                    if (channel != null) {
+                        // Channel already exists, returning it
```
Comment (imakunin) [inline lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/services/ChannelService.java:88]: add (debug) log here, plz

File: lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/services/ChannelService.java [hunk 19]
```diff
@@ -0,0 +1,149 @@
...
+        }
+    }
+
+    @Override
+    public void destroy(LCMPS.DestroyRequest request, StreamObserver<LCMPS.DestroyResponse> responseObserver) {
+        var logPrefix = "(Destroy: {channelId: %s}): ".formatted(request.getChannelId());
+        LOG.info("{} Destroying channel with reason {}", logPrefix, request.getReason());
+
+        try {
+            DbHelper.withRetries(LOG, () -> channelDao.drop(request.getChannelId(), null));
```
Comment (imakunin) [inline lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/services/ChannelService.java:120]: add comment here why we don't send any terminate messages to the peers

File: lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/services/SlotsService.java [hunk 20]
```diff
@@ -0,0 +1,496 @@
...
+        var role = switch (request.getRole()) {
+            case UNSPECIFIED, UNRECOGNIZED -> {
+                LOG.error("{} Cannot determine role", logPrefix);
+
+                throw Status.INVALID_ARGUMENT
+                    .withDescription("Role of peer not set or unsupported")
+                    .asRuntimeException();
+            }
+            case CONSUMER -> CONSUMER;
+            case PRODUCER -> Peer.Role.PRODUCER;
```
Comment (imakunin) [inline lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/services/SlotsService.java:98]: s/Peer.Role.PRODUCER/PRODUCER

File: lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/services/SlotsService.java [hunk 21]
```diff
@@ -0,0 +1,496 @@
...
+            return;
+        }
+
+        final ProducerAndConsumerTransfers res = createProducer(channelId, peerId, peerDesc, logPrefix, role);
+
+        // Finding storage consumer
+        PeerAndTransfer storageConsumer = null;
+        for (var consumer: res.transfers) {
+            if (consumer.peer().description().hasStoragePeer()) {
+                storageConsumer = consumer;
```
Comment (imakunin) [inline lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/services/SlotsService.java:123]: `break` is missing

File: lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/services/SlotsService.java [hunk 22]
```diff
@@ -0,0 +1,496 @@
...
+        } catch (Exception e) {
+            LOG.error("{} Cannot save peer description in db: ", logPrefix, e);
+            throw Status.INTERNAL
+                .withDescription("Cannot save peer description in db")
+                .asRuntimeException();
+        }
+        return result;
+    }
+
+    private record PeerAndTransfer(Peer peer, String transferId) {}
```
Comment (imakunin) [inline lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/services/SlotsService.java:228]: let's format records in the unified format: each field on a separate line

File: lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/services/SlotsService.java [hunk 23]
```diff
@@ -0,0 +1,496 @@
...
+
+    private record ProducerAndConsumerTransfers(Peer producer, List<PeerAndTransfer> transfers) {}
+
+    @Override
+    public void unbind(LCMS.UnbindRequest request, StreamObserver<LCMS.UnbindResponse> responseObserver) {
```
Comment (imakunin) [inline lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/services/SlotsService.java:233]: it's better to order methods by visibility: i.e. `public` methods should be placed before the `private` ones

**Pull request merged**